Internalnotification Plugin

waht opened the pull request:
PR for internalnotification plugin (see #825)

markus2330 commented:
add to whitelist

markus2330 commented on src/plugins/README.md:
```diff
@@ -165,6 +165,10 @@ Log/Send out all changes to configuration to:
 
+Notification of key changes:
+
+- [internalnotification](internalnotification/) automatically update integer variables when a key is changed from inside an application
```
keep the short description more generic. I think "automatically update registered variables"

markus2330 commented on src/plugins/internalnotification/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+ENDIF (NOT PLUGIN_INTERNALNOTIFICATION_VERBOSE)
+
+MARK_AS_ADVANCED(
+	PLUGIN_INTERNALNOTIFICATION_VERBOSE
```
try to avoid the variable, simple use VERBOSE set in kdbconfig.h.
Improvements of Elektra's "logger" are very welcomed.

markus2330 commented on src/plugins/internalnotification/README.md:
```diff
@@ -0,0 +1,37 @@
+
+## Usage ##
+
+Allows programs to register integer variables that get automatically updated
```
Do not mention integer here, it is already mentioned in Todo/Issues

markus2330 commented on src/plugins/internalnotification/README.md:
```diff
@@ -0,0 +1,37 @@
+when the value of the key changes.
+
+Use as global plugin:
+$ sudo kdb global-mount internalnotification
```
Please also give a code example how to use it.

markus2330 commented on src/plugins/internalnotification/README.md:
```diff
@@ -0,0 +1,37 @@
+
+## Exported Methods ##
+
+This plugin exports the function `int elektraInternalnotificationRegisterInt (Plugin * handle, int * variable, Key * key)`.
```
How would a user get a handle to the plugin? You should write a wrapper that takes KDB and does the lookup to your plugin's handle.

markus2330 commented on src/plugins/internalnotification/README.md:
```diff
@@ -0,0 +1,37 @@
+
+## Todo/Issues ##
+
+- Correct data storage for registered keys? elektraPlugin(Set|Get)Data?
```
Isn't this done already?

markus2330 commented on src/plugins/internalnotification/README.md:
```diff
@@ -0,0 +1,37 @@
+## Todo/Issues ##
+
+- Correct data storage for registered keys? elektraPlugin(Set|Get)Data?
+- Can we get rid of exporting the plugin handle?
```
Yes. The idea was that the user calls an initialization function, this function registers itself in _KDB->globalPlugins. That is also the place where you can lookup the handle.

So the user links against your plugin, global mounting of it is not required then.

markus2330 commented on src/plugins/internalnotification/README.md:
```diff
@@ -0,0 +1,37 @@
+- Can we get rid of exporting the plugin handle?
+- Different data types for values:
+  - integer: Before conversion: check if key is string, if key is binary, just use the value
+  - string: Update string value
```
binary data is not so important, string would be useful though

markus2330 commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -0,0 +1,263 @@
+	return item;
+}
+
+static void elektraInternalnotificationUpdateRegisteredKeys (ListPointer * listPointer, KeySet * keySet)
```
Please add comments to functions.

KurtMi replied:
The parameter `listPointer` is not necessary you could call `elektraPluginGetData` in the function.

markus2330 commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -0,0 +1,263 @@
+		}
+		elektraPluginSetData (handle, listPointer);
+
+		// Inizialize list pointers for registered keys
```
typo

markus2330 commented:
Good job, looks very nice! Tests should be fine but documentation needs some improvements! It is important that you explain from end user perspective how to use it.

KurtMi commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -0,0 +1,263 @@
+
+static KeyRegistration * elektraInternalnotificationAddNewRegistration (ListPointer * listPointer)
+{
+	KeyRegistration * item = elektraMalloc (sizeof *item);
```
I think `sizeof (item)` should be enough. With the `*` you allocate more memory then needed.

Namoshek replied:
Is this even legal to reference the variable on the right side during the declaration? I wonder how this works...

tom-wa replied:
it's legal, but better stop talking about it, otherwise you could start a war ;)

KurtMi commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -0,0 +1,263 @@
+struct _ListPointer
+{
+	KeyRegistration * head;
+	KeyRegistration * tail;
```
I do not see any advantage of a tail in an single linked list :)

sanssecours replied:
A better name for `tail` might be `last`. The tail of a list is usually everything except for the first element:

![listmonster](https://cloud.githubusercontent.com/assets/691989/17204483/169cd90e-54a6-11e6-917b-7b1834b4dc2d.png)

Image Source: http://learnyouahaskell.com/starting-out#an-intro-to-lists

KurtMi commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -0,0 +1,263 @@
+
+		// Free list pointer
+		elektraFree (listPointer);
+		elektraPluginSetData (handle, NULL);
```
I think this it not necessary.

tom-wa replied:
free only frees the allocated memory, but the pointer still holds the address. i haven't read the whole code, but it's cleaner and allows you to make checks like elektraPluginGetData == NULL

sanssecours commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -0,0 +1,263 @@
+
+	if (listPointer->head == NULL)
+	{
+		// Inizialize list
```
> // **Inizialize** list

Typo: **Initialize** list

sanssecours commented on src/plugins/internalnotification/README.md:
```diff
@@ -0,0 +1,37 @@
+## Usage ##
+
+Allows programs to register integer variables that get automatically updated
+when the value of the key changes.
```
> when the value of **the** key changes.

Maybe better: when the value of **a** key changes.

Namoshek replied:
That would imply that the integer gets also updated when any other key (and not the one for the integer) in the database changes. I haven't read through everything yet, but I doubt this is the case.

sanssecours replied:
> That would imply that the integer gets also updated when any other key (and not the one for the integer) in the database changes.

You are right. Some might read it that way. Then maybe something like this instead: “…when the value of **a specified** key changes.”

Namoshek commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -0,0 +1,263 @@
+	fprintf (stderr, "elektraInternalnotificationGet keyName=%s\n", parentKeyName);
+#endif
+
+	ListPointer * listPointer = elektraPluginGetData (handle);
```
I think you should check that the fetched data is not NULL and then cast properly (occurs several times).

waht replied:
I added assertions to make sure that listPointer (now pluginState) is not NULL i.e. that the plugin state is initialized.
Normally Elekta's core will call `elektraInternalnotificationOpen` on `kdbOpen` before any other plugin functions. An application would have to call the functions directly in which case the resulting behaviour is unspecified anyway.

waht commented:
@markus2330 it should be ready to merge. please review

markus2330 commented:
@haraldg Do you have some time for a code review?

haraldg commented:
I have little time until the cold season starts, but I'll at least try to understand the API/usage:

So at startup an application registers its int values with elektraInternalnotificationRegisterInt and later whenever kdbGet() is called from the application, the plugin updates these int values? - Note that I needed about an hour to come up with this theory; maybe the documentation not only of the plugin but also of electra itself can be improved, to give an overview of the API more easily.

In case the above theory is correct: Is kdbGet() a cheap operation, if no key has been changed? Is there some way to figure out efficiently when to call kdbGet()?

I don't quite understand why this manual attempt to dynamic linking is necessary, but since it seems to be just a temporary solution, I didn't bother too much to figure it out ... ;-)

markus2330 commented:
@haraldg Thanks for looking into it! Feel free to also add comments to the code if you have some time.

> So at startup an application registers its int values with elektraInternalnotificationRegisterInt and later whenever kdbGet() is called from the application, the plugin updates these int values?

Yes, exactly.

> Note that I needed about an hour to come up with this theory

@waht Can you add a sentence like this in the intro, please? Giving a big picture is always important ;)

> In case the above theory is correct: Is kdbGet() a cheap operation

Yes, kdbGet() is a relatively cheap operation if nothing is to be done (then it only does `stat` on config files to check for changes). We already had kdbGet() polling in some threads/processes on an "embedded" system (I think it was an ARM Cortex-A5) without noticable overhead.

> Is there some way to figure out efficiently when to call kdbGet()?

Yes, via inter-process notification @waht will also work on that (also with event/main loop integration). We are currently thinking about dbus and redis. dbus is in openwrt (and your elektra plugin for it, too). Are you okay with dbus? Do you have some specific requirements?

markus2330 commented on src/plugins/internalnotification/README.md:
```diff
@@ -0,0 +1,92 @@
+## Usage ##
+
+Allows applications to automatically update registered variables when the value of a specified key has changed.
+Support for callbacks that fire when a specified key has changed is planned.
```
Move this to TODOs/Limitations. Please give some big picture in the beginning and then describe things that work.

markus2330 commented on src/plugins/internalnotification/README.md:
```diff
@@ -0,0 +1,92 @@
+This plugin exports the function `int elektraInternalnotificationRegisterInt (Plugin * handle, int * variable, Key * key)`.
+If the given key is contained in a KeySet on a kdbGet or kdbSet operation, its value is
+converted from string to integer and the registered variable is updated with the current value.
+This also works if the registered key is a cascading key.
```
What happens with cascading keys?

markus2330 commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -0,0 +1,312 @@
+		if (pluginState == NULL)
+		{
+#if PLUGIN_INTERNALNOTIFICATION_VERBOSE
+			fprintf (stderr, "elektraInternalnotificationOpen elektraMalloc failed!\n");
```
please use logging also here and you should set an error for outside world. (Which does the logging already for you)

markus2330 commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -0,0 +1,312 @@
+				ELEKTRA_LOG ("conversion failed! keyString=\"%s\" *end=%c, errno=%d, value=%ld", keyString (key), *end,
+					     errno, value);
+			}
+		}
```
log if not found?

markus2330 commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -0,0 +1,312 @@
+			}
+			else
+			{
+				ELEKTRA_LOG ("conversion failed! keyString=\"%s\" *end=%c, errno=%d, value=%ld", keyString (key), *end,
```
Some higher log level here (or even better, add an warning), the user did something wrong here.

haraldg commented:
> Yes, kdbGet() is a relatively cheap operation if nothing is to be done (then it only does stat on config files to check for changes). We already had kdbGet() polling in some threads/processes on an "embedded" system (I think it was an ARM Cortex-A5) without noticable overhead.

My systems currently don't run on battery but I like too keep the option. So I try keep the code efficient even if it isn't strictly necessary ATM. A stat every few seconds is fine of course.

>>    Is there some way to figure out efficiently when to call kdbGet()?

>Yes, via inter-process notification @waht will also work on that (also with event/main loop integration). We are currently thinking about dbus and redis. dbus is in openwrt (and your elektra plugin for it, too). Are you okay with dbus? Do you have some specific requirements?

I'd rather do some slow polling (if it only triggers one stat in the usual case) than depend on dbus. If electra is going to provide some nice way to register the files it is monitoring with an event loop (I'm using libev), this would be even better of course.

markus2330 commented:
> If electra is going to provide some nice way to register the files it is monitoring with an event loop (I'm using libev), this would be even better of course.

Getting all file names and doing inotify on them is possible but that would be rather a workaround because you might get many events even though it is only one `kdbSet`. So better is to fire events within `kdbSet`.

Because any process may trigger changes (using `kdbSet`) and many applications might be interested, we need some kind of a bus system which is not a good idea to reinvent/reimplement. Is there any dep (providing a message bus) that would be okay for you? Usually the message busses return a file descriptor to be integrated in event loops (like libev). 0mq (zmq) would be another candidate next to dbus.

haraldg commented:
> Getting all file names and doing inotify on them is possible but that would be rather a workaround because you might get many events even though it is only one kdbSet.

If each event just triggers one kdbGet() and only the first kdbGet() is actually expensive, then this doesn't sound too bad. OTOH I'm not familiar with the efficiency of the inotify system, maybe it isn't that much better then polling (also given that it has lot's of limitations anyway). It definitely isn't worth your time to implement this, if I would be the only user.

> Is there any dep (providing a message bus) that would be okay for you?

I don't think so. I'm very reluctant to add new dependencies and ATM I don't see enough benefits for doing so.

markus2330 commented:
Ok, thanks for the input! So we will also look into inotify-based and polling mechanism (without deps). They also have the advantage that they take modifications bypassing Elektra into account. I will discuss it with @waht and we will inform you when we find a generic solution (otherwise we will advise you how it can be done for your specific case).

markus2330 commented:
Please rebase, there is a conflict in a file.

If its ready for review, please tag it accordingly.

waht commented:
I added the notification library (wrapper for internalnotification plugin) to this PR. Other notable changes:

- The list plugin allows to pass a plugin handle in its configuration. If a handle is present it uses the plugin instead of loading it. This enables the notification library to load the plugin itself and have a reference to it for getting the functions exported by the plugin.
- I added two fields to the `KDB` struct: `ioBinding` and `notificationPlugin`.

Since the notification library depends on the internalnotification plugin it is excluded from building it the plugin is not included.

markus2330 commented:
Thank you for continuing the work!

> The list plugin has a new configuration key user/plugins/#/handle for passing plugin handles

Would be great if this can be done automatically. Why not use a cascading lookup to allow any namespace? Do we need to make sure that different plugins get different handles? (Some user docu would be great to better understand the purpose.)

waht commented:
The notification library uses the new configuration option to insert the internalnotification plugin automatically. It changes the configuration only at run-time, not permanently.

I confused the naming with the scheme used for global plugins. The option is only called `plugins/#/handle` without "user". I added the option to the readme of the list plugin. Usage of the library is  covered in the notification tutorial. But I could add a readme for the library documenting mounting.

markus2330 commented on src/include/kdbnotification.h:
```diff
@@ -1,7 +1,37 @@
+ *
+ * @defgroup kdbnotification Notification
+ *
+ * @par Global Mounting
```
Thanks for writing the docu.

For uniform style we should put the definition of the group in a comment block after the inclusion of header files, see e.g. in src/libs/elektra/kdb.c.

If there are reasons for your style (e.g. can the `@file` be within a group? Is this useful?) please say so, we might change the other files then.

waht replied:
Sure, I changed that.

Yes, a file can be in a group with the `@ingroup` command. Files are then listed in a Doxygen module page before functions - I used this for the I/O and Notification pages.

waht commented:
I changed the docs and added notifications to the release news. 

@markus2330 I think the PR is ready for review :)

markus2330 commented on src/libs/notification/example/example_notification.c:
```diff
@@ -0,0 +1,105 @@
+	}
+
+	int value = 0;
+	Key * valueKey = keyNew ("/sw/tests/example_notification/#0/current/value", KEY_END);
```
valueKey is an irritating name. What about keyToWatch?

markus2330 commented on src/libs/notification/example/example_notification.c:
```diff
@@ -0,0 +1,105 @@
+	Key * key = keyNew ("/sw/tests/example_notification/#0/current", KEY_END);
+	KDB * kdb = kdbOpen (key);
+	ELEKTRA_NOT_NULL (kdb);
+	if (kdb == NULL)
```
Remove assert because it is checked afterwards anyway. Also is one header file less.

markus2330 commented on src/libs/notification/example/example_notification.c:
```diff
@@ -0,0 +1,105 @@
+
+	printf ("Heyo! My value is %d\n", value);
+	resetTerminalColor ();
+}
```
I doubt that valgrind does not find any memleak here.

waht replied:
Well true, I did not check the example for memory leaks, only the tests :)

markus2330 commented on src/libs/notification/example/example_notification.c:
```diff
@@ -0,0 +1,105 @@
+	ksRewind (ks);
+	while ((cur = ksNext (ks)) != NULL)
+	{
+		printf ("debugKeySet %s = %s\n", keyName (cur), keyString (cur));
```
Please only print relevant two keys?

markus2330 commented on src/libs/notification/example/example_notification.c:
```diff
@@ -0,0 +1,105 @@
+#include <string.h>
+
+// from https://en.wikipedia.org/wiki/ANSI_escape_code#Colors
+#define ANSI_COLOR_RESET "\x1b[0m"
```
Are they used in this example?

waht replied:
Yes the are used to set the colors in `setTerminalColor()`.

markus2330 replied:
Okay, it was not clear to me how to use this then.

markus2330 commented on src/libs/notification/example/example_notification.c:
```diff
@@ -0,0 +1,105 @@
+	kdbGet (kdb, config, key);
+	debugKeySet (config);
+
+	printf ("Heyo! My value is %d\n", value);
```
The example works great but without loop this is not really visible!

waht replied:
Great! I will add a loop.

waht commented:
Great! I extended the example and added change detection to the internalnotification plugin. For string keys conversion or callbacks are now only triggered if the key's value has changed. Binary keys are triggered at every kdbGet.

markus2330 commented:
Looks good now, is this ready to merge?

markus2330 commented:
jenkins build all please

waht commented:
jenkins build all please

waht commented:
Yes I think so.

About the failed build checks: 
- "elektra-mergerequests-fast": seems to had an unclean cmake env (there is no typechecker plugin in my branch) 
- "elektra-multiconfig-gcc-stable": the test "testshell_markdown_blockresolver" failed - should be unrelated
- "elektra-multiconfig-gcc47-cmake-options": the test "testshell_hosts" failed - should also be unrelated 
- @sanssecours is already on the build issue on Travis/OSX (#1817)

markus2330 commented:
jenkins build all please

markus2330 commented:
I triggered all build jobs, let us hope they succeed now.

markus2330 commented:
I am afraid you need to rebase so that travis succeeds again. See #1817

waht commented:
jenkins build all please

markus2330 commented:
Thank you for creating this plugin!
